src/QsCompiler/CompilationManager/CompilationUnit.cs

```diff
@@ -883,14 +889,32 @@ public QsCompilation Build()
                 // having duplicate names in the syntax tree.
                 var (taggedCallables, taggedTypes) = RenameInternalDeclarations(callables, types, predicate: source => this.Externals.Declarations.ContainsKey(source));
                 var tree = NewSyntaxTree(taggedCallables, taggedTypes, this.GlobalSymbols.Documentation());
-                return new QsCompilation(tree, entryPoints.ToImmutable());
+
+                if (this.CachedCallGraph.IsNull)
+                {
+                    this.CachedCallGraph = QsNullable<ICallGraph>.NewValue(Transformations.CallGraphWalker.BuildCallGraph.Apply(taggedCallables));
+                }
+
+                return new QsCompilation(tree, entryPoints.ToImmutable(), this.CachedCallGraph);
             }
             finally
             {
                 this.syncRoot.ExitReadLock();
             }
         }
 
+        /// <summary>
+        /// Checks the cycles of the given callables and returns diagnostics for
+        /// any invalid cycles, along with the qualified name of the callable the
+        /// diagnostic should be placed.
+        /// </summary>
+        public IEnumerable<Tuple<QsQualifiedName, QsCompilerDiagnostic>> VerifyCycles(IEnumerable<QsCallable> callables)
+        {
+            this.CachedCallGraph = QsNullable<ICallGraph>.NewValue(Transformations.CallGraphWalker.BuildCallGraph.Apply(callables));
```

Comment:
Does this always need to rebuild the call graph, or can it reuse a cached version if it exists?

Reply:
The arguments to this method are the callables that are being verified against, so if someone calls this method with several callables, but there is already a cached call graph, we would't want to ignore the given callable in favor of the old call graph. Here we want to make a new call graph with the given callable, set it as the cache and verify if the cycles in the new graph are valid.

Reply:
What happens in this situation:

1. `VerifyCycles` is called with a list of callables that is not the same as `this.GlobalSymbols.DefinedCallables()`. It sets the cached call graph.
2. `Build` is called afterwards. It checks `this.CachedCallGraph.IsNull` and does not regenerate the call graph.

Does the compilation returned by `Build` have an incorrect call graph?

Reply:
If the callables given to VerifyCycles are not the same callables that are used in the Build() method, than the call graph would not be accurate to the returned syntax tree.

Reply:
I think without looking at the implementation of both `VerifyCycles` and `Build`, it is not obvious that that sequence of events causes `Build` to behave incorrectly, since `VerifyCycles` does not indicate that it changes the state of the compilation unit. It seems like it would be better to prevent this scenario from happening.

Reply:
This method has been removed because the QsCompilation no longer has a call graph field. RunTypeChecking will generate its own call graph for generating diagnostics.